### PR TITLE
Allow user access to errors from Spectrum

### PR DESCRIPTION
**Pull request by CiaranH1997**
Adds new functions to Hist and Spectrum classes. These allow the user to directly access the errors from Spectrum without having to plot a ROOT histogram.

### CODE REVIEW

File: CAFAna/Core/Spectrum.cxx [hunk 1]
```diff
@@ -269,6 +269,15 @@ namespace ana
       return (exposure/fLivetime) * fHist.GetEigenStan();
   }
 
+  //----------------------------------------------------------------------
+  Eigen::ArrayXd Spectrum::GetEigenSqErrors(double exposure, EExposureType expotype) const
+  {
+    if(expotype == kPOT)
+      return std::pow(exposure/fPOT, 2) * fHist.GetEigenSqErrors();
+    else
+      return std::pow(exposure/fLivetime, 2) * fHist.GetEigenSqErrors();
```
Comment (cjbackhouse) [inline CAFAna/Core/Spectrum.cxx:278]:
```suggestion
    if(expotype == kPOT)
      return util::sqr(exposure/fPOT) * fHist.GetEigenSqErrors();
    else
      return util::sqr(exposure/fLivetime) * fHist.GetEigenSqErrors();
```

**Comment by cjbackhouse** [comment]
Rescuing some discussion from slack

> [Chris Backhouse]
> What happens if you call GetEigenSqErrors() on a Hist that only has fSqrtErrs errors? You get an empty error array back, right? > That seems suboptimal
> I think if the user cares enough about errors to be calling this, it's probably OK to make the errors concrete by filling in the array with the sqrt-based errors
> that should maybe respect the env var though, not sure
>
> [Ciaran Hasnip] Ahh okay, so a change could be if (!fSumSqErrs) return fSumSq; else { fill fSumSq with sqrt of fData and return this; }
>
> [Chris Backhouse] and I guess the else would set fSumSqErrs=false;

@CiaranH1997 if you make that change I am happy to accept this PR

File: CAFAna/Core/Hist.h [hunk 2]
```diff
@@ -45,6 +45,15 @@ namespace ana
...
     const Eigen::ArrayXd& GetEigen() const {assert(fType == kDense); return fData;}
     const Eigen::ArrayXstan& GetEigenStan() const {assert(fType == kDenseStan); return fDataStan;}
+    const Eigen::ArrayXd& GetEigenSqErrors() const {
+      assert(fType == kDense);
+      if (fSqrtErrs) {
+        fSumSq = fData.sqrt();
+        return fSumSq;
+      } else {
+        return fSumSq;
+      }
```
Comment (cjbackhouse) [inline CAFAna/Core/Hist.h:55]:
```suggestion
      if(fSqrtErrs){
        fSumSq = fData.sqrt();
        fSqrtErrs = false;
      }
      return fSumSq;
    }
```

File: CAFAna/Core/Hist.h [hunk 3]
```diff
@@ -45,6 +45,15 @@ namespace ana
...
     const Eigen::ArrayXstan& GetEigenStan() const {assert(fType == kDenseStan); return fDataStan;}
+    const Eigen::ArrayXd& GetEigenSqErrors() const {
+      assert(fType == kDense);
+      if(fSqrtErrs){
+        fSumSq = fData.sqrt();
+        fSqrtErrs = false;
+      }
+      return fSumSq;
+    }
+    }
```
Comment (cjbackhouse) [inline CAFAna/Core/Hist.h:56]:
```suggestion
    const Eigen::ArrayXd& GetEigenSqErrors() const
    {
      assert(fType == kDense);
      if(fSqrtErrs){
        fSumSq = fData.sqrt();
        fSqrtErrs = false;
      }
      return fSumSq;
    }
```

**Comment by cjbackhouse** [comment]
Thanks. I fiddled around a bit more, and merged.

Do you need a release with this in, or is it OK if it just shows up the next time something else needs to be updated?

**Comment by CiaranH1997** [comment]
Okay great thanks!

There is no rush to release this as I don't think it is critical. Happy to wait until the next update comes!

Cheers,
Ciaran
________________________________

**Pull request merged**